lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Logical/Impl/FlatTopology.cs

```diff
@@ -0,0 +1,429 @@
+
+            lock (_lock)
+            {
+                if (!_nodes.ContainsKey(id))
+                {
+                    throw new ArgumentException("Task is not part of this topology");
+                }
+
+                DataNode node = _nodes[id];
+                var prevState = node.FailState;
+                node.FailState = DataNodeState.Lost;
+                _nodesWaitingToJoinTopologyNextIteration.Remove(taskId);
+                _nodesWaitingToJoinTopology.Remove(taskId);
+                _lostNodesToBeRemoved.Add(taskId);
+
+                if (prevState != DataNodeState.Reachable)
+                {
+                    return 0;
+                }
+
+                _availableDataPoints--;
+            }
+
+            return 1;
+        }
+
+        /// <summary>
+        /// Whether the topology can be sheduled.
+        /// </summary>
+        /// <returns>True if the topology is ready to be scheduled</returns>
+        public bool CanBeScheduled()
+        {
+            return _root != null;
+        }
+
+        /// <summary>
+        /// Finalizes the topology.
+        /// After the topology has been finalized, any task added to the topology is
+        /// assumed as a task recovered from a failure.
+        /// </summary>
+        /// <returns>The same finalized topology</returns>
+        public ITopology Build()
+        {
+            if (_finalized == true)
+            {
+                throw new IllegalStateException("Topology cannot be built more than once");
+            }
+
+            if (_root == null)
+            {
+                throw new IllegalStateException("Topology cannot be built because the root node is missing");
+            }
+
+            if (OperatorId <= 0)
+            {
+                throw new IllegalStateException("Topology cannot be built because not linked to any operator");
+            }
+
+            if (StageName == string.Empty)
+            {
+                throw new IllegalStateException("Topology cannot be built because not linked to any stage");
+            }
+
+            BuildTopology();
+
+            _rootTaskId = Utils.BuildTaskId(_taskStage, _rootId);
+            _finalized = true;
+
+            return this;
+        }
+
+        /// <summary>
+        /// Utility method for logging the topology state.
+        /// This will be called every time a topology object is built or modified
+        /// because of a failure.
+        /// </summary>
+        public string LogTopologyState()
+        {
+            var children = _root.Children.GetEnumerator();
+            string output = _rootId + "\n";
+            while (children.MoveNext())
+            {
+                var rep = "X";
+                if (children.Current.FailState == DataNodeState.Reachable)
+                {
+                    rep = children.Current.TaskId.ToString();
+                }
+
+                output += rep + " ";
+            }
+
+            return output;
```

Comment:
I dunno. The failing node should be already present in the log because the task manager logs it once it receives a failure. This is more sort of like "hey we got a failure, this is the new state of world".